Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/Saml2AllowNoSubjectAssertionSpecValidator.java

```diff
@@ -43,32 +49,65 @@ public Saml2AllowNoSubjectAssertionSpecValidator() {
         super();
     }
 
+    /*
+     * (non-Javadoc)
+     *
+     * @see
+     * gov.hhs.fha.nhinc.callback.cxf.Saml2ExchangeAuthFrameworkValidator#validateAssertion(org.apache.wss4j.common.saml
+     * .SamlAssertionWrapper)
+     */
+    @Override
+    protected void validateAssertion(final SamlAssertionWrapper samlAssertion) throws WSSecurityException {
+        try {
+            validateSubject(samlAssertion.getSaml2());
+        } catch (final WSSecurityException e) {
+            LOG.error("Unable to validate Assertion ", e);
+            throw e;
+        }
+
+    }
+
     /**
      * Checks that the Subject element is present when required, but does not check for the existence of Subject when
      * the Assertion contains AttributeStatements.
      *
      * @param assertion the assertion
      * @throws ValidationException the validation exception
      */
+
     @Override
-    protected void validateSubject(Assertion assertion) throws ValidationException {
-        if ((assertion.getStatements() == null || assertion.getStatements().isEmpty())
-                && (assertion.getAuthnStatements() == null || assertion.getAuthnStatements().isEmpty())
-                && (assertion.getAttributeStatements() == null || assertion.getAttributeStatements().isEmpty())
-                && (assertion.getAuthzDecisionStatements() == null || assertion.getAuthzDecisionStatements().isEmpty())
-                && assertion.getSubject() == null) {
-            throw new ValidationException("Subject is required when Statements are absent");
-        }
+    protected void validateSubject(final Assertion assertion) throws WSSecurityException {
 
-        if (assertion.getAuthnStatements().size() > 0 && assertion.getSubject() == null) {
-            throw new ValidationException("Assertions containing AuthnStatements require a Subject");
+        if (CollectionUtils.isEmpty(assertion.getStatements())) {
+            assertAttAuthzStatements(assertion);
+        }
+        if (CollectionUtils.isNotEmpty(assertion.getAuthnStatements()) && assertion.getSubject() == null) {
+            throw new WSSecurityException(ErrorCode.FAILURE, "Assertions containing AuthnStatements require a Subject");
         }
-        if (assertion.getAuthzDecisionStatements().size() > 0 && assertion.getSubject() == null) {
-            throw new ValidationException("Assertions containing AuthzDecisionStatements require a Subject");
+        if (CollectionUtils.isNotEmpty(assertion.getAuthzDecisionStatements()) && assertion.getSubject() == null) {
+            throw new WSSecurityException(ErrorCode.FAILURE,
+                "Assertions containing AuthzDecisionStatements require a Subject");
         }
 
         if (assertion.getSubject() != null) {
             validateSubject(assertion.getSubject());
         }
     }
-}
\ No newline at end of file
+
+    protected static void assertAttAuthzStatements(final Assertion assertion) throws WSSecurityException {
```

Comment:
I don't know if other devs have the same issue, but I find it tough to follow a lot of the SAML logic that's happening here -- not because of the code (that part's clean and easy to trace through), but because of my lack of understanding of the workflow / functional logic behind it.  Consider adding some very brief Javadocs (or inline comments) to make this easier to follow.